### PR TITLE
feat: SITES-9481 Initial implementation

### PR DESCRIPTION
**Pull request by tyge68**
Please ensure your pull request adheres to the following guidelines:
- [X] make sure to link the related issues in this description
- [X] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues

see SITES-9481 (internal)

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+} from '@aws-sdk/client-s3';
+
+const VALID_MODES = ['preview', 'live'];
+const VALID_ACTIONS = ['store', 'evict'];
+
+const buildDefaultParams = () => ({
+  Bucket: 'franklin-content-bus-headless',
+});
+
+const deleteItemAction = (s3, key, deletedKeys) => new Promise((resolve, reject) => {
```
Comment (tripodsan) [inline src/index.js:27]: why `Promise()` and not just `async` ?

File: src/index.js [hunk 2]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+  const params = buildDefaultParams();
+  params.Key = key;
```
Comment (tripodsan) [inline src/index.js:29]: this is kind of a weird pattern. better:

```suggestion
  const params = buildDefaultParams({
    Key: key,
  });
```

File: src/index.js [hunk 3]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+  s3.send(new DeleteObjectCommand(params)).then(
+    (_) => {
+      deletedKeys.push(params.Key);
```
Comment (tripodsan) [inline src/index.js:32]: this is ugly, adding the Key to a result array. why not just return it?

File: src/index.js [hunk 4]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+      resolve();
+    },
+  ).catch((err) => {
+    reject(err);
+  });
+});
+
+const storeInPreview = async (s3, payload, options) => {
```
Comment (tripodsan) [inline src/index.js:40]: we prefer `function` style over the arrow pattern.

File: src/index.js [hunk 5]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+  };
+  try {
+    await s3.send(new PutObjectCommand(params));
+    return new Response(`${params.Key} stored in S3 bucket`);
+  } catch (err) {
+    return new Response(`An error occurred while trying to store ${params.Key} in S3 bucket due to ${err.message}`, { status: 500 });
+  }
+};
+
+const copyPreviewToLive = async (s3, options) => {
```
Comment (tripodsan) [inline src/index.js:60]: we prefer `function` style over the arrow pattern.

File: src/index.js [hunk 6]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+  const params = buildDefaultParams();
+  params.CopySource = encodeURI(`${params.Bucket}/${options.s3SourceObjectPath}${options.suffix}.json`);
```
Comment (tripodsan) [inline src/index.js:62]: are you sure this needs to be encoded ?

File: src/index.js [hunk 7]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+  params.Key = `${options.s3TargetObjectPath}${options.suffix}.json`;
+  try {
+    await s3.send(new CopyObjectCommand(params));
+    return new Response(`${params.Key} copy preview to live in S3 bucket`);
+  } catch (err) {
+    return new Response(`An error occurred while trying to copy ${params.CopySource}  to ${params.Key} in S3 bucket due to ${err.message}`, { status: 500 });
+  }
+};
+
+const evictFromS3 = async (s3, s3ObjectPath) => {
```
Comment (tripodsan) [inline src/index.js:72]: we prefer `function` style over the arrow pattern.

File: src/index.js [hunk 8]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+  try {
+    const params = buildDefaultParams();
+    params.Prefix = `${s3ObjectPath}.`;
+    const listResponse = await s3.send(new ListObjectsCommand(params));
```
Comment (tripodsan) [inline src/index.js:76]: this is not sufficient, the ListObjectCommand might only return the default paging size object.
see https://github.com/adobe/helix-admin/blob/965d46b14e33207f810db438dd534145481b43a4/src/support/storage.js#L347-L368 for an example for reading all results.

File: src/index.js [hunk 9]
```diff
@@ -13,7 +13,83 @@ import wrap from '@adobe/helix-shared-wrap';
...
+    delete params.Prefix;
+    const deletedKeys = [];
+    const promises = [];
+    for (const item of listResponse.Contents) {
+      promises.push(deleteItemAction(s3, item.Key, deletedKeys));
+    }
+    await Promise.all(promises);
```
Comment (tripodsan) [inline src/index.js:83]: better use https://github.com/adobe/helix-shared/tree/main/packages/helix-shared-process-queue

File: src/index.js [hunk 10]
```diff
@@ -22,19 +98,64 @@ import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
...
+      }
+      const mode = json.mode || 'preview';
+      if (!VALID_MODES.includes(mode)) {
+        return new Response(`Invalid parameters mode value, accept:${VALID_MODES}`, { status: 400 });
+      }
+      const action = json.action || 'store';
+      if (!VALID_ACTIONS.includes(action)) {
+        return new Response(`Invalid parameters action value, accept:${VALID_ACTIONS}`, { status: 400 });
+      }
+      const variation = json.variation || 'master';
```
Comment (tripodsan) [inline src/index.js:125]: `master` is a politically incorrect terminology. use `main` instead.

File: src/index.js (hunk 6 above)
Comment (tripodsan) [reply src/index.js:62]: ah right: https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/copyobjectcommandinput.html#copysource

File: src/index.js (hunk 10 above)
Comment (tyge68) [reply src/index.js:125]: sorry but content fragment still use master as the naming convention, so main would be wrong here

File: src/storage.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+  return new Promise((resolve, reject) => {
+    zlib.gzip(str, (err, buffer) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(buffer);
+      }
+    });
+  });
+}
```
Comment (tripodsan) [inline src/storage.js:32]: simpler to use promisify:

```js
import { promisify } from 'util';
import zlib from 'zlib';
.
.
const gzip = promisify(zlib.gzip);
const gunzip = promisify(zlib.gunzip);
```

File: src/storage.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+export default class Storage {
+  constructor() {
+    this.s3 = new S3Client();
+    this.bucket = 'franklin-content-bus-headless';
+  }
+
+  buildDefaultParams(options) {
```
Comment (tripodsan) [inline src/storage.js:40]:
```suggestion
  buildDefaultParams(options = {}) {
```

**Review by tripodsan — CHANGES_REQUESTED** [review]
LGTM, only some minor details.

**Review by tripodsan — APPROVED** [review]
perfect. thanks!

**Pull request merged**